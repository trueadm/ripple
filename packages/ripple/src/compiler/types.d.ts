import * as a from '#acorn';

export type DeclarationKind = 'let' | 'const' | 'var' | 'import' | 'function' | 'component' | 'rest_param' | 'param' | 'using' | 'await using';
export type BindingKind = 'normal';

// I think this stuff comes from svelte, should it get removed at some point?
export type LetDirective = unknown;
export type Binding = { kind: BindingKind; [key: string]: unknown; };
