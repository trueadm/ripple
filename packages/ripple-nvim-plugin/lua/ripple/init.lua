local M = {}

---Entrypoint invoked by plugin managers.
---@param opts? {treesitter?: table|boolean, lsp?: table|boolean}
function M.setup(opts)
	opts = opts or {treesitter = { ensure_installed = true }, lsp = {}}

	if opts.treesitter ~= false then
		require("ripple.treesitter").setup(opts.treesitter or {})
	end

	if opts.lsp ~= false then
		require("ripple.lsp").setup(opts.lsp or {})
	end
end

return M
