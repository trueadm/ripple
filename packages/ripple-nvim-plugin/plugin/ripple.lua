if vim.filetype and vim.filetype.add then
	vim.filetype.add({
		extension = {
			ripple = "ripple",
		},
	})
end
