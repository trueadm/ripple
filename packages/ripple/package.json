{
  "name": "ripple",
  "description": "Ripple is an elegant TypeScript UI framework",
  "license": "MIT",
  "author": "Dominic Gannaway",
  "version": "0.2.67",
  "type": "module",
  "module": "src/runtime/index.js",
  "main": "src/runtime/index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/trueadm/ripple.git",
    "directory": "packages/ripple"
  },
  "bugs": {
    "url": "https://github.com/trueadm/ripple/issues"
  },
  "homepage": "https://ripplejs.com",
  "keywords": [
    "ripple",
    "UI",
    "framework"
  ],
  "exports": {
    ".": {
      "types": "./types/index.d.ts",
      "worker": "./src/runtime/index.js",
      "browser": "./src/runtime/index.js",
      "default": "./src/runtime/index.js"
    },
    "./package.json": "./package.json",
    "./compiler": {
      "types": "./types/index.d.ts",
      "require": "./compiler/index.js",
      "default": "./src/compiler/index.js"
    },
    "./validator": {
      "types": "./types/index.d.ts",
      "require": "./validator/index.js",
      "default": "./src/validator/index.js"
    },
    "./internal/client": {
      "default": "./src/runtime/internal/client/index.js"
    },
    "./jsx-runtime": {
      "types": "./src/jsx-runtime.d.ts",
      "import": "./src/jsx-runtime.js",
      "default": "./src/jsx-runtime.js"
    }
  },
  "imports": {
    "#client": "./src/runtime/internal/client/types.d.ts",
    "#client/constants": "./src/internal/client/constants.js"
  },
  "dependencies": {
    "@jridgewell/sourcemap-codec": "^1.5.5",
    "acorn": "^8.15.0",
    "acorn-typescript": "^1.4.13",
    "esrap": "^2.1.0",
    "is-reference": "^3.0.3",
    "magic-string": "^0.30.18",
    "muggle-string": "^0.4.1",
    "zimmerframe": "^1.1.2"
  },
  "devDependencies": {
    "@types/estree": "^1.0.8",
    "@types/node": "^24.3.0",
    "typescript": "catalog:"
  }
}
