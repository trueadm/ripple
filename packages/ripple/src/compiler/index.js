/** @import { RawSourceMap } from 'source-map' */
/** @import { Program } from 'estree' */
/** @import { ParseError } from './phases/1-parse/index.js' */

import { parse as parse_module } from './phases/1-parse/index.js';
import { analyze } from './phases/2-analyze/index.js';
import { transform_client } from './phases/3-transform/client/index.js';
import { transform_server } from './phases/3-transform/server/index.js';
import { convert_source_map_to_mappings } from './phases/3-transform/segments.js';

/**
 * @param {string} source 
 * @returns {{ ast: Program, errors: ParseError[] }}
 */
export function parse(source) {
  return parse_module(source);
}

/**
 * @param {string} source 
 * @param {string} filename 
 * @param {{ mode?: 'client' | 'server' }} options 
 * @returns {{ js: { code: string, map: RawSourceMap }, css: { code: string, map: RawSourceMap } | null }}
 */
export function compile(source, filename, options = {}) {
  const ast = parse_module(source);
  const analysis = analyze(ast, filename);
  const result = options.mode === 'server'
      ? transform_server(filename, source, analysis)
      : transform_client(filename, source, analysis, false);

  return result;
}

export function compile_to_volar_mappings(source, filename) {
  // Parse and transform to get the esrap sourcemap
  const ast = parse_module(source);
  const analysis = analyze(ast, filename);
  const transformed = transform_client(filename, source, analysis, true);

  return convert_source_map_to_mappings(transformed.js.map, source, transformed.js.code);
}
