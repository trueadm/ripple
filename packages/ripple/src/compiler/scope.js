/** @import * as ESTree from 'estree' */

import is_reference from 'is-reference';
import { extract_identifiers, object, unwrap_pattern } from '../utils/ast.js';
import { walk } from 'zimmerframe';
import { is_reserved } from './utils.js';
import * as b from '../utils/builders.js';

export function create_scopes(ast, root, parent) {
  const scopes = new Map();
  const scope = new Scope(root, parent, false);
  scopes.set(ast, scope);

  const state = { scope };
  const references = [];
  const updates = [];

  function add_params(scope, params) {
    for (const param of params) {
      for (const node of extract_identifiers(param)) {
        scope.declare(node, 'normal', param.type === 'RestElement' ? 'rest_param' : 'param');
      }
    }
  }

  const create_block_scope = (node, { state, next }) => {
    const scope = state.scope.child(true);
    scopes.set(node, scope);

    if (node.type === 'ForOfStatement') {
      if (node.index) {
        state.scope.declare(node.index, 'normal', 'let');
      }
    }

    next({ scope });
  };

  walk(ast, state, {
    // references
    Identifier(node, { path, state }) {
      const parent = path.at(-1);
      if (
        parent &&
        is_reference(node, /** @type {Node} */ (parent)) &&
        // TSTypeAnnotation, TSInterfaceDeclaration etc - these are normally already filtered out,
        // but for the migration they aren't, so we need to filter them out here
        // TODO -> once migration script is gone we can remove this check
        !parent.type.startsWith('TS')
      ) {
        references.push([state.scope, { node, path: path.slice() }]);
      }
    },

    AssignmentExpression(node, { state, next }) {
      updates.push([state.scope, node.left]);
      next();
    },

    UpdateExpression(node, { state, next }) {
      updates.push([state.scope, /** @type {Identifier | MemberExpression} */ (node.argument)]);
      next();
    },

    ImportDeclaration(node, { state }) {
      for (const specifier of node.specifiers) {
        state.scope.declare(specifier.local, 'normal', 'import', node);
      }
    },

    Component(node, { state, next }) {
      const scope = state.scope.child();
      scopes.set(node, scope);

      if (node.id) scope.declare(node.id, 'normal', 'component');

      add_params(scope, node.params);
      next({ scope });
    },

    Element(node, { state, next }) {
      const scope = state.scope.child();
      scopes.set(node, scope);

      next({ scope });
    },

    FunctionExpression(node, { state, next }) {
      const scope = state.scope.child();
      scopes.set(node, scope);

      if (node.id) scope.declare(node.id, 'normal', 'function');

      add_params(scope, node.params);
      next({ scope });
    },

    FunctionDeclaration(node, { state, next }) {
      if (node.id) state.scope.declare(node.id, 'normal', 'function', node);

      const scope = state.scope.child();
      scopes.set(node, scope);

      add_params(scope, node.params);
      next({ scope });
    },

    ArrowFunctionExpression(node, { state, next }) {
      const scope = state.scope.child();
      scopes.set(node, scope);

      add_params(scope, node.params);
      next({ scope });
    },

    ForStatement: create_block_scope,
    ForInStatement: create_block_scope,
    ForOfStatement: create_block_scope,
    SwitchStatement: create_block_scope,
    BlockStatement(node, context) {
      const parent = context.path.at(-1);
      if (
        parent?.type === 'FunctionDeclaration' ||
        parent?.type === 'FunctionExpression' ||
        parent?.type === 'ArrowFunctionExpression'
      ) {
        // We already created a new scope for the function
        context.next();
      } else {
        create_block_scope(node, context);
      }
    },

    Eval(node, context) {
      const finalizer = node.finalizer;

      if (finalizer !== null) {
        for (const node of finalizer.body) {
          context.visit(node);
        }
      }
    },

    ClassDeclaration(node, { state, next }) {
      if (node.id) state.scope.declare(node.id, 'normal', 'let', node);
      next();
    },

    VariableDeclaration(node, { state, path, next }) {
      for (const declarator of node.declarations) {
        /** @type {Binding[]} */
        const bindings = [];

        state.scope.declarators.set(declarator, bindings);

        for (const id of extract_identifiers(declarator.id)) {
          const binding = state.scope.declare(id, 'normal', node.kind, declarator.init);
          bindings.push(binding);
        }
      }

      next();
    },

    CatchClause(node, { state, next }) {
      if (node.param) {
        const scope = state.scope.child(true);
        scopes.set(node, scope);

        for (const id of extract_identifiers(node.param)) {
          scope.declare(id, 'normal', 'let');
        }

        next({ scope });
      } else {
        next();
      }
    },
  });

  for (const [scope, { node, path }] of references) {
    scope.reference(node, path);
  }

  for (const [scope, node] of updates) {
    for (const expression of unwrap_pattern(node)) {
      const left = object(expression);
      const binding = left && scope.get(left.name);

      if (binding !== null && left !== binding.node) {
        binding.updated = true;

        if (left === expression) {
          binding.reassigned = true;
        } else {
          binding.mutated = true;
        }
      }
    }
  }

  return {
    scope,
    scopes,
  };
}

export class Scope {
  /** @type {ScopeRoot} */
  root;

  /**
   * The immediate parent scope
   * @type {Scope | null}
   */
  parent;

  /**
   * Whether or not `var` declarations are contained by this scope
   * @type {boolean}
   */
  #porous;

  /**
   * A map of every identifier declared by this scope, and all the
   * identifiers that reference it
   * @type {Map<string, Binding>}
   */
  declarations = new Map();

  /**
   * A map of declarators to the bindings they declare
   * @type {Map<VariableDeclarator | AST.LetDirective, Binding[]>}
   */
  declarators = new Map();

  /**
   * A set of all the names referenced with this scope
   * â€” useful for generating unique names
   * @type {Map<string, { node: Identifier; path: AST.Node[] }[]>}
   */
  references = new Map();

  /**
   * The scope depth allows us to determine if a state variable is referenced in its own scope,
   * which is usually an error. Block statements do not increase this value
   */
  function_depth = 0;

  /**
   * If tracing of reactive dependencies is enabled for this scope
   * @type {null | Expression}
   */
  tracing = null;

  /**
   *
   * @param {ScopeRoot} root
   * @param {Scope | null} parent
   * @param {boolean} porous
   */
  constructor(root, parent, porous) {
    this.root = root;
    this.parent = parent;
    this.#porous = porous;
    this.function_depth = parent ? parent.function_depth + (porous ? 0 : 1) : 0;
  }

  /**
   * @param {Identifier} node
   * @param {Binding['kind']} kind
   * @param {DeclarationKind} declaration_kind
   * @param {null | Expression | FunctionDeclaration | ClassDeclaration | ImportDeclaration} initial
   * @returns {Binding}
   */
  declare(node, kind, declaration_kind, initial = null) {
    if (this.parent) {
      if (declaration_kind === 'var' && this.#porous) {
        return this.parent.declare(node, kind, declaration_kind);
      }

      if (declaration_kind === 'import') {
        return this.parent.declare(node, kind, declaration_kind, initial);
      }
    }

    if (node.name === '_$_') {
      throw new Error('Cannot declare a variable named "_$_" as it is a reserved identifier');
    }

    if (this.declarations.has(node.name)) {
      throw new Error(`'${node.name}' has already been declared in the current scope`);
    }

    /** @type {Binding} */
    const binding = {
      node,
      references: [],
      initial,
      reassigned: false,
      mutated: false,
      updated: false,
      scope: this,
      kind,
      declaration_kind,
      is_called: false,
      metadata: null,
    };

    this.declarations.set(node.name, binding);
    this.root.conflicts.add(node.name);
    return binding;
  }

  child(porous = false) {
    return new Scope(this.root, this, porous);
  }

  /**
   * @param {string} preferred_name
   * @returns {string}
   */
  generate(preferred_name) {
    if (this.#porous) {
      return /** @type {Scope} */ (this.parent).generate(preferred_name);
    }

    preferred_name = preferred_name.replace(/[^a-zA-Z0-9_$]/g, '_').replace(/^[0-9]/, '_');
    let name = preferred_name;
    let n = 1;

    while (
      this.references.has(name) ||
      this.declarations.has(name) ||
      this.root.conflicts.has(name) ||
      is_reserved(name)
    ) {
      name = `${preferred_name}_${n++}`;
    }

    this.references.set(name, []);
    this.root.conflicts.add(name);
    return name;
  }

  /**
   * @param {string} name
   * @returns {Binding | null}
   */
  get(name) {
    return this.declarations.get(name) ?? this.parent?.get(name) ?? null;
  }

  /**
   * @param {VariableDeclarator | AST.LetDirective} node
   * @returns {Binding[]}
   */
  get_bindings(node) {
    const bindings = this.declarators.get(node);
    if (!bindings) {
      throw new Error('No binding found for declarator');
    }
    return bindings;
  }

  /**
   * @param {string} name
   * @returns {Scope | null}
   */
  owner(name) {
    return this.declarations.has(name) ? this : this.parent && this.parent.owner(name);
  }

  /**
   * @param {Identifier} node
   * @param {AST.Node[]} path
   */
  reference(node, path) {
    path = [...path]; // ensure that mutations to path afterwards don't affect this reference
    let references = this.references.get(node.name);

    if (!references) this.references.set(node.name, (references = []));

    references.push({ node, path });

    const binding = this.declarations.get(node.name);
    if (binding) {
      binding.references.push({ node, path });
    } else if (this.parent) {
      this.parent.reference(node, path);
    } else {
      // no binding was found, and this is the top level scope,
      // which means this is a global
      this.root.conflicts.add(node.name);
    }
  }
}

export class ScopeRoot {
  /** @type {Set<string>} */
  conflicts = new Set();

  /**
   * @param {string} preferred_name
   */
  unique(preferred_name) {
    preferred_name = preferred_name.replace(/[^a-zA-Z0-9_$]/g, '_');
    let final_name = preferred_name;
    let n = 1;

    while (this.conflicts.has(final_name)) {
      final_name = `${preferred_name}_${n++}`;
    }

    this.conflicts.add(final_name);
    const id = b.id(final_name);
    return id;
  }
}
