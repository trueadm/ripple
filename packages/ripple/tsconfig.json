{
	"compilerOptions": {
		"module": "esnext",
		"lib": ["esnext", "dom", "dom.iterable"],
		"target": "esnext",
		"noEmit": true,
		"moduleResolution": "bundler",
		"resolveJsonModule": true,
		"noEmitOnError": true,
		"noErrorTruncation": true,
		"allowSyntheticDefaultImports": true,
		"verbatimModuleSyntax": true,
		"types": ["node"],
		"jsx": "preserve",
		"jsxImportSource": "ripple",
		"strict": true,
		"allowJs": true,
		"checkJs": true,
		"paths": {
			"ripple": ["./types/index.d.ts"],
		}
	},
	"include": [
		"./*.js",
		"./src/",
		"./tests/*/test.ripple",
	]
}
