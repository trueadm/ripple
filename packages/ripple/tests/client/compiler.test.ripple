import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { mount, TrackedArray, track } from 'ripple';
import { parse } from 'ripple/compiler'

describe('compiler success tests', () => {
  let container;

	function render(component) {
		mount(component, {
			target: container
		});
	}

	beforeEach(() => {
		container = document.createElement('div');
		document.body.appendChild(container);
	});

	afterEach(() => {
		document.body.removeChild(container);
		container = null;
	});


	it('Parses style content correctly', () => {
		const source = `export component App() {
  <div id="myid" class="myclass">{"Hello World"}</div>

  <style>#style</style>
}`;
		const style1 = '.myid {color: green }';
		const style2 = '#myid {color: green }';
		const style3 = 'div {color: green }';

		let input = source.replace('#style', style1);
		let ast = parse(input);
		expect(ast.body[0].declaration.css.source).toEqual(style1);

		input = source.replace('#style', style2);
		ast = parse(input);
		expect(ast.body[0].declaration.css.source).toEqual(style2);

		input = source.replace('#style', style3);
		ast = parse(input);
		expect(ast.body[0].declaration.css.source).toEqual(style3);
	});

  it('renders without crashing', () => {
		component App() {
			let foo;
			let bar;
			let baz;

			foo = {};
			foo = {'test': 0};
			foo['abc'] = 123;

			bar = { 'def': 456 };

			baz = { 'ghi': 789 };
			baz['jkl'] = 987;
		}

		render(App);
	});


  	it('renders without crashing using < character', () => {
		component App() {
			function bar() {
        for (let i = 0; i < 10; i++) {
          // do nothing
        }
        const x = 1 < 1;
      }

			let x = 5 < 10

      <div>{x}</div>
		}

		render(App);
	});

	it('render lexical blocks without crashing', () => {
		component App() {
			<div>
				const a = 1;
				<div>
					const b = 1;
				</div>
				<div>
					const b = 1;
				</div>
			</div>
			<div>
				const a = 2;
				<div>
					const b = 1;
				</div>
			</div>
		}

		render(App);
	});

	it('properly handles JS assignments, reads and updates to array indices', () => {
		const logs = [];

		component App() {
			let items = [];
			let tracked_items = track([]);
			let items2 = new Array();
			let items3 = new TrackedArray();
			let i = 0;

			logs.push(items[0]);
			logs.push(items[i]);
			logs.push(@tracked_items[0]);
			logs.push(@tracked_items[i]);
			logs.push(items2[0]);
			logs.push(items2[i]);
			logs.push(items3[0]);
			logs.push(items3[i]);

			items[0] = 123;
			items[i] = 123;
			@tracked_items[0] = 123;
			@tracked_items[i] = 123;
			items2[0] = 123;
			items2[i] = 123;
			items3[0] = 123;
			items3[i] = 123;

			logs.push(items[0]);
			logs.push(items[i]);
			logs.push(@tracked_items[0]);
			logs.push(@tracked_items[i]);
			logs.push(items2[0]);
			logs.push(items2[i]);
			logs.push(items3[0]);
			logs.push(items3[i]);

			items[0]++;
			items[i]++;
			@tracked_items[0]++;
			@tracked_items[i]++;
			items2[0]++;
			items2[i]++;
			items3[0]++;
			items3[i]++;

			logs.push(items[0]);
			logs.push(items[i]);
			logs.push(@tracked_items[0]);
			logs.push(@tracked_items[i]);
			logs.push(items2[0]);
			logs.push(items2[i]);
			logs.push(items3[0]);
			logs.push(items3[i]);

			logs.push(--items[0]);
			logs.push(--items[i]);
			logs.push(--@tracked_items[0]);
			logs.push(--@tracked_items[i]);
			logs.push(--items2[0]);
			logs.push(--items2[i]);
			logs.push(--items3[0]);
			logs.push(--items3[i]);
		}

		render(App);

		expect(logs).toEqual([
			undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined,
			123, 123, 123, 123, 123, 123, 123, 123,
			125, 125, 125, 125, 125, 125, 125, 125,
			124, 123, 124, 123, 124, 123, 124, 123
		]);
	});

	it('renders without crashing using mapped types', () => {
		component App() {
			type RecordKey = 'test';
			type RecordValue = { a: string, b: number };

			const config: Record<RecordKey, RecordValue> = {
				test: {
					a: 'test',
					b: 1
				},
			};

			const config2: { [key in RecordKey]: RecordValue } = {
				test: {
					a: 'test2',
					b: 2
				}
			}

			const config3: { [key: RecordKey]: RecordValue } = {
				test: {
					a: 'test3',
					b: 3
				}
			}
		}

		render(App);
	});

	it('renders without crashing using object destructuring', () => {
		component App() {
			const obj = { a: 1, b: 2, c: 3 };
			const { a, b, ...rest } = obj;

			<div>
				{'a '}{a} {'b '} {b} {'rest '} {JSON.stringify(rest)}

				<div>

				</div>
			</div>
		}

		render(App);
	});

	it('renders without crashing using object destructuring #2', () => {
		component App() {
			const obj = { a: 1, b: 2, c: 3 };
			const { a, b, ...rest } = obj;

			{'a '}{a} {'b '} {b} {'rest '} {JSON.stringify(rest)}

			<div>

			</div>
		}

		render(App);
	});

	it('should not fail with random TS syntax', () => {
		function tagFn() {
			return null;
		}

		function Wrapper() {
		return {
				unwrap: function<T>() {
				return null as unknown as T;
				}
			}
		}

		component App() {
			let x: number[] = [] as number[];

			const n = new Wrapper<number>().unwrap<string>();

			const tagResult = tagFn`value`;

			interface Node<T> {
				value: T;
			}

			class Box<T> {
				value: T;

				method<T>(): T {
				return this.value;
				}
			}

			let flag = true;

			const s = flag ? new Box<number>() : new Box<string>();
		}

		render(App);
	});
});
