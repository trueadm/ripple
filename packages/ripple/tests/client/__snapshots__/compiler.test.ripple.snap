// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler success tests > compiles TSInstantiationExpression 1`] = `
"import * as _$_ from 'ripple/internal/client';

function makeBox(value) {
	return { value };
}

const makeStringBox = (makeBox);
const stringBox = makeStringBox('abc');
const ErrorMap = (Map);
const errorMap = new ErrorMap();"
`;

exports[`compiler success tests > compiles tracked values in effect with assignment expression 1`] = `"state.count = _$_.get(count);"`;

exports[`compiler success tests > compiles tracked values in effect with update expressions 1`] = `
"_$_.with_scope(__block, () => untrack(() => {
			state.preIncrement = _$_.update_pre(count, __block);
			state.postIncrement = _$_.update(count, __block);
			state.preDecrement = _$_.update_pre(count, __block, -1);
			state.postDecrement = _$_.update(count, __block, -1);
		}));"
`;
